### ISSUE TITLE
Design: Add optional `locale` argument to *localSubdomain* calls?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#11982]
As a Chapel programmer, I'd like the ability to not only reason about what sublocale(s) of a domain/array my current locale (`here`) owns, but also what other locales own.  To that end, this issue proposes adding an optional `loc = here` argument to the three current *localSubdomain* calls in order to support such queries without adding new routines.  Because this argument is optional, existing codes should work without change while new codes can make use of the new feature.  

This interface is being proposed in order to address issue #10405.

**Comment by lydia-duncan** [comment]
Would this function be able to promote naturally over an array of locales?

**Comment by bradcray** [comment]
> Would this function be able to promote naturally over an array of locales?

Presumably, though that wasn't inherent in my goal in proposing it.  Do you see any reason that it wouldn't be able to be?

**Comment by cassella** [comment]
Since `hasSingleLocalSubdomain` is a param function, wouldn't its `loc` need to be a param?

**Comment by bradcray** [comment]
> Since hasSingleLocalSubdomain is a param function, wouldn't its loc need to be a param?

Not as currently defined and written.  Currently that is a property of the distribution (think of it as "has the potential to have multiple sublocales"), not of the actual execution-time locales as well (since not even the current implied `here` value is a `param`.

Oh, but, intentionally or not you've pointed out a stupid oversight in my proposal (and the draft implementation I put together) which is that the `loc` argument wouldn't actually be used in that routine, so should be omitted, duh.

While pursuing this prototype, I also realized that the `param`-ness of this function is a little questionable for other reasons (see issue #11930).

More generally, param functions can take non-param arguments... they just can't do anything with their execution time values, only with their static type information.  For example, you could write:

```chapel
proc numBits(x) param {
  return ...;
}
```

where `x` could be an arbitrary variable rather than its type (say).

**Comment by mppf** [comment]
👍

**Comment by lydia-duncan** [comment]
> >    Would this function be able to promote naturally over an array of locales?

> Presumably, though that wasn't inherent in my goal in proposing it. Do you see any reason that it wouldn't be able to be?

No, but I also don't know the rules about promotion very well.  I'm satisfied that you think it would, but would want a test to lock in that case to go with the implementation.
